src/librustc/util/profiling.rs

```diff
@@ -35,409 +40,125 @@ pub enum ProfilerEvent {
     QueryBlockedEnd { query_name: &'static str, category: ProfileCategory, time: u64 },
 }
 
-impl ProfilerEvent {
-    fn timestamp(&self) -> u64 {
-        use self::ProfilerEvent::*;
-
-        match self {
-            QueryStart { time, .. } |
-            QueryEnd { time, .. } |
-            GenericActivityStart { time, .. } |
-            GenericActivityEnd { time, .. } |
-            QueryCacheHit { time, .. } |
-            QueryCount { time, .. } |
-            IncrementalLoadResultStart { time, .. } |
-            IncrementalLoadResultEnd { time, .. } |
-            QueryBlockedStart { time, .. } |
-            QueryBlockedEnd { time, .. } => *time
-        }
-    }
-}
-
 fn thread_id_to_u64(tid: ThreadId) -> u64 {
     unsafe { mem::transmute::<ThreadId, u64>(tid) }
 }
 
 pub struct SelfProfiler {
-    events: FxHashMap<ThreadId, Vec<ProfilerEvent>>,
-    start_time: SystemTime,
-    start_instant: Instant,
+    profiler: Profiler,
+    query_event_kind: StringId,
+    generic_activity_event_kind: StringId,
+    incremental_load_result_event_kind: StringId,
+    query_blocked_event_kind: StringId,
+    query_cache_hit_event_kind: StringId,
 }
 
 impl SelfProfiler {
     pub fn new() -> SelfProfiler {
-        let profiler = SelfProfiler {
-            events: Default::default(),
-            start_time: SystemTime::now(),
-            start_instant: Instant::now(),
+        let filename = format!("pid-{}.rustc_profile", process::id());
+        let path = std::path::Path::new(&filename);
+        let profiler = Profiler::new(path);
+
+        let query_event_kind = profiler.alloc_string("Query");
+        let generic_activity_event_kind = profiler.alloc_string("GenericActivity");
+        let incremental_load_result_event_kind = profiler.alloc_string("IncrementalLoadResult");
+        let query_blocked_event_kind = profiler.alloc_string("QueryBlocked");
+        let query_cache_hit_event_kind = profiler.alloc_string("QueryCacheHit");
+
+        SelfProfiler {
+            profiler,
+            query_event_kind,
+            generic_activity_event_kind,
+            incremental_load_result_event_kind,
+            query_blocked_event_kind,
+            query_cache_hit_event_kind,
+        }
+    }
+
+    fn get_query_name_string_id(query_name: QueryName) -> StringId {
+        let discriminant = unsafe {
+            mem::transmute::<_, u64>(mem::discriminant(&query_name))
```

Comment:
Maybe make this `mem::transmute::<Discriminant<QueryName>, u64>(mem::discriminant(&query_name))` in order for it to be a little more safe.